### ISSUE TITLE
Ambiguous call when overloads are constrained with complementary constraints involving dependent NTTP

### ISSUE THREAD

**Issue by ahatanak** [llvm/llvm-project#196375]
clang incorrectly reports an ambiguous call when two function template overloads are constrained with `requires C` and `requires (!C)`, where the concept `C` depends on a non-type template parameter whose type uses `__underlying_type(En)`.

$ cat test.cpp
```c++
template <typename T>
concept Enum = __is_enum(T);
enum NamedEnum { A, B, C };

template <typename En, __underlying_type(En) value>
concept Small = (value <= 2);

static_assert(!Small<NamedEnum, 4>); // OK

template <Enum En, __underlying_type(En) value>
consteval bool f()
requires Small<En, value>
{ 
  return true;
} 
template <Enum En, __underlying_type(En) value>
consteval bool f()
requires(!Small<En, value>)
{ 
  return false;
} 

static_assert(f<NamedEnum, 1>() == true);  // OK
static_assert(f<NamedEnum, 4>() == false); // ERROR: ambiguous
```

$ clang++ -std=c++23 -c test.cpp
```
  error: call to 'f' is ambiguous
     24 | static_assert(f<NamedEnum, 4>() == false); // ERROR: ambiguous
        |               ^~~~~~~~~~~~~~~
  note: candidate function [with En = NamedEnum, value = 4]
     11 | consteval bool f()
        |                ^
  note: candidate function [with En = NamedEnum, value = 4]
     17 | consteval bool f()
        |                ^
```

**Comment by frederick-vs-ja** [comment]
Also reproducible in C++20 mode ([Godbolt link](https://godbolt.org/z/qPPc79bxr)).

**Comment by cor3ntin** [comment]
Further reduced

```cpp
template <typename En, En value>
concept Small = (value <= 2);

template <int value>
consteval bool f()
    requires(Small<int, value>
             // && value == 1
    )
{
    return true;
}
template <int value>
consteval bool f()
    requires(!Small<int, value>)
{
    return false;
}

static_assert(f<4>() == false);
```

**Comment by cor3ntin** [comment]
Interestingly, this is fine

```
template <int value>
concept Small = (value <= 2);

template <int value>
consteval bool f()
    requires(Small<value>
             // && value == 1
    )
{
    return true;
}
template <int value>
consteval bool f()
    requires(!Small<value>)
{
    return false;
}

static_assert(f<4>() == false);
```

maybe we are not correctly substituting the typename into the constant template argument?

**Comment by cor3ntin** [comment]
I'm an idiot. This reduces to 

```cpp
template <typename En, En value>
concept Small = (value <= 2);

template <int value>
consteval bool f()
    requires(Small<int, value>
             // && value == 1
    )
{
    return true;
}
static_assert(f<4>());
```

**Comment by cor3ntin** [comment]
We end up creating a `SubstNonTypeTemplateParmExpr` for `value`. Later, we try to rewrite that expression but clamg assumes we are inside an instantiation  - however we never create instantiations for concepts - it's unclear what the consequences of creating `ImplicitConceptSpecializationDecl` would be - or how we would map then to a concept + args.

So we fail to substitute `value` properly - and it ends up dependent - we treat dependent concept as being satisfied - arguably the fact we fail to surface the error is itself a bug,

@mizvekov @zyn0217
